File: server/proliferate/server/cloud/workflows/contracts/legacy_upgrade.py
"""Deterministic UUIDv5 identity derivation for the legacy-definition upgrade
(feature spec §5.1).

Legacy definitions without identities are upgraded by a new immutable version
whose UUIDv5 IDs are derived once from the fixed Proliferate namespace. The old
version is retained for audit and never rewritten.
"""

from __future__ import annotations

import uuid

# Fixed Proliferate workflow-identity namespace (feature spec §5.1).
PROLIFERATE_WORKFLOW_NAMESPACE = uuid.UUID("2b5e907a-2cd8-5b8f-b5ab-5c891bb93263")

_KINDS = frozenset({"slot", "node", "group", "lane", "step"})


def legacy_identity_name(workflow_version_id: str, kind: str, identity: str) -> str:
    """The exact UTF-8 UUIDv5 name string (feature spec §5.1)."""

    if kind not in _KINDS:
        raise ValueError(f"unknown identity kind {kind!r}")
    return f"workflow-version={workflow_version_id}\nkind={kind}\nidentity={identity}"


def derive_legacy_id(workflow_version_id: str, kind: str, identity: str) -> str:
    """Derive the deterministic lowercase UUIDv5 for a legacy structural object."""

    name = legacy_identity_name(workflow_version_id, kind, identity)
    return str(uuid.uuid5(PROLIFERATE_WORKFLOW_NAMESPACE, name))

File: apps/packages/product-domain/src/workflows/contracts/legacy-upgrade.ts
/**
 * Deterministic UUIDv5 identity derivation for the legacy-definition upgrade
 * (feature spec §5.1). Matches the Python `legacy_upgrade.py` output.
 */

import { sha1Bytes, utf8Bytes } from "./hashing";

/** Fixed Proliferate workflow-identity namespace (feature spec §5.1). */
export const PROLIFERATE_WORKFLOW_NAMESPACE = "2b5e907a-2cd8-5b8f-b5ab-5c891bb93263";

export type LegacyIdentityKind = "slot" | "node" | "group" | "lane" | "step";

const KINDS: ReadonlySet<string> = new Set(["slot", "node", "group", "lane", "step"]);

export function legacyIdentityName(
  workflowVersionId: string,
  kind: LegacyIdentityKind,
  identity: string,
): string {
  if (!KINDS.has(kind)) {
    throw new Error(`unknown identity kind ${kind}`);
  }
  return `workflow-version=${workflowVersionId}\nkind=${kind}\nidentity=${identity}`;
}

function uuidBytes(uuid: string): Uint8Array {
  const hex = uuid.replace(/-/g, "");
  const out = new Uint8Array(16);
  for (let i = 0; i < 16; i += 1) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function formatUuid(bytes: Uint8Array): string {
  let hex = "";
  for (const b of bytes) {
    hex += b.toString(16).padStart(2, "0");
  }
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

function uuid5(namespace: string, name: string): string {
  const ns = uuidBytes(namespace);
  const nameBytes = utf8Bytes(name);
  const combined = new Uint8Array(ns.length + nameBytes.length);
  combined.set(ns);
  combined.set(nameBytes, ns.length);
  const hash = sha1Bytes(combined);
  const out = hash.slice(0, 16);
  out[6] = (out[6] & 0x0f) | 0x50; // version 5
  out[8] = (out[8] & 0x3f) | 0x80; // RFC 4122 variant
  return formatUuid(out);
}

export function deriveLegacyId(
  workflowVersionId: string,
  kind: LegacyIdentityKind,
  identity: string,
): string {
  return uuid5(PROLIFERATE_WORKFLOW_NAMESPACE, legacyIdentityName(workflowVersionId, kind, identity));
}
